File: app/src/main/java/com/ducluanxutrieu/quanlynhanvien/Activity/ChatsActivity.java
package com.ducluanxutrieu.quanlynhanvien.Activity;

import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;

import com.ducluanxutrieu.quanlynhanvien.Adapter.MessageAdapter;
import com.ducluanxutrieu.quanlynhanvien.Models.Friend;
import com.ducluanxutrieu.quanlynhanvien.Models.MessageItem;
import com.ducluanxutrieu.quanlynhanvien.MyTask;
import com.ducluanxutrieu.quanlynhanvien.R;
import com.ducluanxutrieu.quanlynhanvien.Models.TokenUser;
import com.firebase.ui.database.FirebaseRecyclerOptions;
import com.google.android.gms.tasks.Continuation;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.functions.FirebaseFunctions;
import com.google.firebase.functions.FirebaseFunctionsException;
import com.google.firebase.functions.HttpsCallableResult;
import com.google.firebase.messaging.FirebaseMessaging;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ChatsActivity extends AppCompatActivity {
    RecyclerView mChatRecyclerView;
    EditText mInputChatEditText;
    ImageButton mSendImageButton;
    MessageAdapter mMessageAdapter;
    List<MessageItem> mMessageItems;

    //Firebase
    FirebaseDatabase mMessageFireDatabase;
    //DatabaseReference mMessageReferenceSend;
    //DatabaseReference mMessageReferenceReceive;
    FirebaseFunctions mFunctions;
    //ChildEventListener mMessageChildEvenListener;
    FirebaseMessaging mFirebaseMessage;
    FirebaseAuth mMessageFireAuth;
    FirebaseUser mUserFire;

    Friend friend;
    String rootName;

    private static String token;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chats);
        mapping();
        mMessageItems = new ArrayList<>();

        Intent intent = getIntent();
        friend = (Friend) intent.getSerializableExtra("friend");
        setTitle(friend.getName());

        mFunctions = FirebaseFunctions.getInstance();
        mMessageFireDatabase = FirebaseDatabase.getInstance();
        mMessageFireAuth = FirebaseAuth.getInstance();
        mFirebaseMessage = FirebaseMessaging.getInstance();
        mUserFire = mMessageFireAuth.getCurrentUser();


        if (mMessageFireAuth.getCurrentUser() != null){
            rootName = mMessageFireAuth.getCurrentUser().getDisplayName();
        }

        mInputChatEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.toString().trim().length() > 0){
                    mSendImageButton.setEnabled(true);
                }else {
                    mSendImageButton.setEnabled(false);
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        mSendImageButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //pushNoti();
                addMessage(mInputChatEditText.getText().toString()).addOnCompleteListener(new OnCompleteListener<String>() {
                    @Override
                    public void onComplete(@NonNull Task<String> task) {
                        if (!task.isSuccessful()) {
                            Exception e = task.getException();
                            if (e instanceof FirebaseFunctionsException) {
                                FirebaseFunctionsException ffe = (FirebaseFunctionsException) e;
                                FirebaseFunctionsException.Code code = ffe.getCode();
                                Object details = ffe.getDetails();
                            }

                            // ...
                        }

                        // ...
                    }
                });
                mInputChatEditText.setText("");
            }
        });

        getToken();
    }

    private Task<String> addMessage(String text) {
        // Create the arguments to the callable function.
        Map<String, Object> data = new HashMap<>();
        data.put("text", text);
        data.put("to", friend.getUid());
        data.put("push", true);

        return mFunctions
                .getHttpsCallable("addMessage")
                .call(data)
                .continueWith(new Continuation<HttpsCallableResult, String>() {
                    @Override
                    public String then(@NonNull Task<HttpsCallableResult> task) throws Exception {
                        // This continuation runs on either success or failure, but if the task
                        // has failed then getResult() will throw an Exception which will be
                        // propagated down.
                        return (String) task.getResult().getData();
                    }
                });
    }


    private void pushNoti() {
        MyTask task = new MyTask(ChatsActivity.this);

        final String BASE_URL = "https://quan-ly-nhan-vien.firebaseapp.com/api";
        //final String BASE_URL = "http://localhost:5000/api";
        String name = rootName;
        String message = mInputChatEditText.getText().toString();
        String to = token;
        String query = "{\"data\":{\"title\":\""+name +"\",\"message\":\""+ message + "\"},\"to\":\""+to+"\"}";
        Log.i("QUERY", query);
        task.execute(BASE_URL, query);
    }

    private void getToken() {
        runtimeEnableAutoInit();
        String email =  friend.getUid();
        String pathChild = "token/" + email;
        DatabaseReference getTokenReference = mMessageFireDatabase.getReference().child(pathChild);
        getTokenReference.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                TokenUser tokenUser = dataSnapshot.getValue(TokenUser.class);
                if (tokenUser == null){
                    token = null;
                }else {
                    token = tokenUser.getTokenU();
                    Log.i("kiemtra1", token);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    public void runtimeEnableAutoInit() {
        // [START fcm_runtime_enable_auto_init]
        FirebaseMessaging.getInstance().setAutoInitEnabled(true);
        // [END fcm_runtime_enable_auto_init]
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
//        mMessageReferenceSend.removeEventListener(mMessageChildEvenListener);
        //mMessageChildEvenListener = null;
        mMessageItems.clear();
    }

    private void mapping(){
        mChatRecyclerView = findViewById(R.id.chats_recycler_view);
        mInputChatEditText = findViewById(R.id.input_chats_edit_text);
        mSendImageButton = findViewById(R.id.send_message_image_button);
    }

    @Override
    protected void onStart() {
        super.onStart();
        Query query = FirebaseDatabase.getInstance().getReference().child("messages" + "/" + mUserFire.getUid() + "/" + friend.getUid()).limitToLast(50);
        FirebaseRecyclerOptions.Builder<MessageItem> itemBuilder = new FirebaseRecyclerOptions.Builder<>();
        itemBuilder.setQuery(query, MessageItem.class);
        FirebaseRecyclerOptions<MessageItem> options = itemBuilder.build();

        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        mChatRecyclerView.setLayoutManager(layoutManager);
        mMessageAdapter = new MessageAdapter(options);
        mMessageAdapter.startListening();
        mChatRecyclerView.setAdapter(mMessageAdapter);
    }

    @Override
    protected void onStop() {
        super.onStop();
        mMessageAdapter.stopListening();
    }
}

File: app/src/main/java/com/ducluanxutrieu/quanlynhanvien/activity/ChatsActivity.kt
package com.ducluanxutrieu.quanlynhanvien.activity

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import android.widget.ImageButton
import android.widget.ProgressBar
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.ducluanxutrieu.quanlynhanvien.adapter.MessageAdapter
import com.ducluanxutrieu.quanlynhanvien.models.MessageItem
import com.ducluanxutrieu.quanlynhanvien.R
import com.firebase.ui.database.FirebaseRecyclerOptions
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.Query
import com.google.firebase.functions.FirebaseFunctions
import com.google.firebase.messaging.FirebaseMessaging
import java.util.*

class ChatsActivity : AppCompatActivity() {
    //recycler
    internal lateinit var layoutManager: LinearLayoutManager
    internal lateinit var mChatRecyclerView: RecyclerView
    private lateinit var query: Query

    private lateinit var mInputChatEditText: EditText
    internal lateinit var mSendImageButton: ImageButton
    internal lateinit var mMessageAdapter: MessageAdapter
    private lateinit var mMessageItems: MutableList<MessageItem>

    //Firebase
    private lateinit var mMessageFireDatabase: FirebaseDatabase
    private lateinit var mFunctions: FirebaseFunctions
    private lateinit var mFirebaseMessage: FirebaseMessaging
    private lateinit var mMessageFireAuth: FirebaseAuth
    private var mUserFire: FirebaseUser? = null

    private var rootName: String? = null
    private var uidFriend: String? = null
    private var nameFriend: String? = null


    //loadmore
    private lateinit var loadmoreProgress: ProgressBar
    private var limitToLoad = 50

    private val timeNow: String
        get() {
            val time: String
            val calendar = Calendar.getInstance()
            val hour = calendar.get(Calendar.HOUR_OF_DAY)
            val minute = calendar.get(Calendar.MINUTE)
            time = "$hour:$minute"
            return time
        }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chats)
        mapping()

        getInstance()

        getRootName()

        getDataFromIntent()

        getChatTextChanged()
    }

    fun getInstance(){
        mMessageItems = ArrayList()

        mFunctions = FirebaseFunctions.getInstance()
        mMessageFireDatabase = FirebaseDatabase.getInstance()
        mMessageFireAuth = FirebaseAuth.getInstance()
        mFirebaseMessage = FirebaseMessaging.getInstance()
        mUserFire = mMessageFireAuth.currentUser
    }

    private fun getRootName() {
        if (mMessageFireAuth.currentUser != null) {
            rootName = mMessageFireAuth.currentUser!!.displayName
        }
    }

    private fun getDataFromIntent() {
        uidFriend = intent.getStringExtra("uid")
        nameFriend = intent.getStringExtra("name")
        title = nameFriend
    }

    fun buttonSendListener(view : View){
        addMessage(mInputChatEditText.text.toString())
        mInputChatEditText.setText("")
    }

    private fun getChatTextChanged() {
        mInputChatEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                mSendImageButton.isEnabled = s.toString().trim { it <= ' ' }.isNotEmpty()
            }

            override fun afterTextChanged(s: Editable) {

            }
        })
    }

    private fun addMessage(text: String): Task<String> {
        // Create the arguments to the callable function.
        val data = HashMap<String, String>()
        data["text"] = text
        data["timeStamp"] = timeNow
        data["to"] = uidFriend!!

        return mFunctions
                .getHttpsCallable("addMessage")
                .call(data)
                .continueWith { task ->
                    // This continuation runs on either success or failure, but if the task
                    // has failed then getResult() will throw an Exception which will be
                    // propagated down.
                    task.result!!.data as String?
                }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        mMessageItems.clear()
    }

    private fun mapping() {
        mChatRecyclerView = findViewById(R.id.chats_recycler_view)
        mInputChatEditText = findViewById(R.id.input_chats_edit_text)
        mSendImageButton = findViewById(R.id.send_message_image_button)
        loadmoreProgress = findViewById(R.id.loadmore_progress)
    }

    override fun onStart() {
        super.onStart()
        query = FirebaseDatabase.getInstance().reference.child("messages" + "/" + mUserFire!!.uid + "/" + uidFriend).limitToLast(limitToLoad)
        val itemBuilder = FirebaseRecyclerOptions.Builder<MessageItem>()
        itemBuilder.setQuery(query, MessageItem::class.java)
        val options = itemBuilder.build()

        layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        mChatRecyclerView.layoutManager = layoutManager
        mMessageAdapter = MessageAdapter(options)
        mMessageAdapter.startListening()
        mChatRecyclerView.adapter = mMessageAdapter

        mMessageAdapter.registerAdapterDataObserver(object : RecyclerView.AdapterDataObserver() {
            override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
                super.onItemRangeInserted(positionStart, itemCount)
                layoutManager.smoothScrollToPosition(mChatRecyclerView, null, mMessageAdapter.itemCount - 1)
            }
        })
    }

    override fun onStop() {
        super.onStop()
        mMessageAdapter.stopListening()
    }
}